[nomerge] Reduce allocations in groupBy

retronym opened the pull request:
- Avoid lambda allocation in getOrElseUpdate call
  - Avoid intermediate Tuple2 instances when iterating through
    the mutable map.
  - Avoid allocating Map1-4 during building the result if
    the size of the grouped exceeds 4. Intead build a HashMap
    directly.

```
./benchdb list --limit 2 && ./benchdb results --run 5 --run 6 --pivot size  && ./benchdb results --run 5 --run 6 --pivot size --metric "·gc.alloc.rate.norm"
┏━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃ ID ┃ Timestamp           ┃ Msg              ┃
┣━━━━╋━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━┫
┃  6 ┃ 2020-08-24 07:14:36 ┃ PR 8948          ┃
┃  5 ┃ 2020-08-24 07:06:47 ┃ PR 8948 Baseline ┃
┗━━━━┻━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━┛
2 test runs found (limit reached).
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━┓
┃                                                (size) ┃                ┃                ┃      ┃     ┃        128         ┃         512         ┃         2048          ┃         8192          ┃       ┃
┃ Benchmark                                             ┃ (hashCodeCost) ┃ (maxNumGroups) ┃ Mode ┃ Cnt ┃    Score ┃   Error ┃     Score ┃   Error ┃      Score ┃    Error ┃      Score ┃    Error ┃ Units ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━╋━━━━━╋━━━━━━━━━━╋━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━┫
┃ #5:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 8631.963 ┃ 116.790 ┃ 30969.939 ┃ 456.711 ┃ 119820.013 ┃ 1741.671 ┃ 480813.306 ┃ 7268.723 ┃ ns/op ┃
┃ #6:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 8287.717 ┃ 135.038 ┃ 30672.551 ┃ 540.409 ┃ 119356.530 ┃ 1592.624 ┃ 477570.346 ┃ 6365.010 ┃ ns/op ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━┻━━━━━━━━━━┻━━━━━━━━━┻━━━━━━━━━━━┻━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┓
┃                                                (size) ┃                ┃                ┃      ┃     ┃       128        ┃        512        ┃        2048        ┃        8192        ┃       ┃
┃ Benchmark [·gc.alloc.rate.norm]                       ┃ (hashCodeCost) ┃ (maxNumGroups) ┃ Mode ┃ Cnt ┃    Score ┃ Error ┃     Score ┃ Error ┃     Score ┃  Error ┃      Score ┃ Error ┃ Units ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━╋━━━━━╋━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━╋━━━━━━━┫
┃ #5:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 2192.014 ┃ 0.022 ┃ 13712.051 ┃ 0.076 ┃ 50949.533 ┃ 56.381 ┃ 198544.803 ┃ 1.213 ┃ B/op  ┃
┃ #6:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 1896.014 ┃ 0.021 ┃  5224.051 ┃ 0.077 ┃ 17768.204 ┃  0.305 ┃  67176.786 ┃ 1.180 ┃ B/op  ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━┻━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┻━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━┻━━━━━━━┛
```



2.12.x appropriate version of of #8947

retronym committed:
Reduce allocations in groupBy

  - Avoid lambda allocation in getOrElseUpdate call
  - Avoid intermediate Tuple2 instances when iterating through
    the mutable map.
  - Avoid allocating Map1-4 during building the result if
    the size of the grouped exceeds 4. Intead build a HashMap
    directly.

```
./benchdb list --limit 2 && ./benchdb results --run 5 --run 6 --pivot size  && ./benchdb results --run 5 --run 6 --pivot size --metric "·gc.alloc.rate.norm"
┏━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃ ID ┃ Timestamp           ┃ Msg              ┃
┣━━━━╋━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━┫
┃  6 ┃ 2020-08-24 07:14:36 ┃ PR 8948          ┃
┃  5 ┃ 2020-08-24 07:06:47 ┃ PR 8948 Baseline ┃
┗━━━━┻━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━┛
2 test runs found (limit reached).
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━┓
┃                                                (size) ┃                ┃                ┃      ┃     ┃        128         ┃         512         ┃         2048          ┃         8192          ┃       ┃
┃ Benchmark                                             ┃ (hashCodeCost) ┃ (maxNumGroups) ┃ Mode ┃ Cnt ┃    Score ┃   Error ┃     Score ┃   Error ┃      Score ┃    Error ┃      Score ┃    Error ┃ Units ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━╋━━━━━╋━━━━━━━━━━╋━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━┫
┃ #5:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 8631.963 ┃ 116.790 ┃ 30969.939 ┃ 456.711 ┃ 119820.013 ┃ 1741.671 ┃ 480813.306 ┃ 7268.723 ┃ ns/op ┃
┃ #6:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 8287.717 ┃ 135.038 ┃ 30672.551 ┃ 540.409 ┃ 119356.530 ┃ 1592.624 ┃ 477570.346 ┃ 6365.010 ┃ ns/op ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━┻━━━━━━━━━━┻━━━━━━━━━┻━━━━━━━━━━━┻━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┓
┃                                                (size) ┃                ┃                ┃      ┃     ┃       128        ┃        512        ┃        2048        ┃        8192        ┃       ┃
┃ Benchmark [·gc.alloc.rate.norm]                       ┃ (hashCodeCost) ┃ (maxNumGroups) ┃ Mode ┃ Cnt ┃    Score ┃ Error ┃     Score ┃ Error ┃     Score ┃  Error ┃      Score ┃ Error ┃ Units ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━╋━━━━━╋━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━╋━━━━━━━┫
┃ #5:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 2192.014 ┃ 0.022 ┃ 13712.051 ┃ 0.076 ┃ 50949.533 ┃ 56.381 ┃ 198544.803 ┃ 1.213 ┃ B/op  ┃
┃ #6:scala.collection.GroupByBenchmark.buildArrayBuffer ┃             32 ┃              8 ┃ avgt ┃  30 ┃ 1896.014 ┃ 0.021 ┃  5224.051 ┃ 0.077 ┃ 17768.204 ┃  0.305 ┃  67176.786 ┃ 1.180 ┃ B/op  ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━┻━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┻━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━┻━━━━━━━┛
```

retronym commented on src/library/scala/collection/TraversableLike.scala:
```diff
@@ -451,17 +452,19 @@ trait TraversableLike[+A, +Repr] extends Any
+      val bldr = m.getOrElse(key, null) match {
+        case null =>
+          val b = newBuilder
+          m.+=((key, b))
```
It is a bit more work, and something of an orthoganal improvement, to avoid the intermediate `Tuple2` here. We need to thread the key, value, and an optional tuple through the creation methods in `HashMap` (ie `addOne(k: K, v: V, kvOrNull: (K, V)`).

Sketch: #8935

This PR stands on its own though.

mkeskells commented on src/library/scala/collection/TraversableLike.scala:
```diff
@@ -451,17 +452,19 @@ trait TraversableLike[+A, +Repr] extends Any
-      b += ((k, v.result))
-
-    b.result
+    m.result(_.result())
```
cant we add a `releaseFence` here only to provide the protection that we need

mkeskells commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -1162,6 +1165,28 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
 
+    def getOrElse[V1 >: B](a: A, f: => V1): V1 = getOrElse0(rootNode, a, computeHash(a), 0, f)
+    def result[V1](f: B => V1): HashMap[A, V1] = {
+      val workList = mutable.ArrayStack[HashMap[A, B]]()
```
we are going to create lots of worklists. Can we just create one in the caller and reuse
Ideally with a Array based srtructure so push doesnt create an object

mkeskells reviewed (commented):
only had a quick look. Will checkout the branch and check again later

retronym committed:
Benchmark for groupBy

retronym commented:
@mkeskells I've reworked this in what I believe is the best approach for 2.12.x and added benchmark results. Care to take a second look?

retronym commented on src/library/scala/collection/TraversableLike.scala:
```diff
@@ -451,17 +451,23 @@ trait TraversableLike[+A, +Repr] extends Any
+        super.entriesIterator
+    }
+    val newBuilderFunction = () => newBuilder
+    for (elem <- this.seq) {
```
The existing code wasn't safe with parallel collections because of the absent `.seq` call and use of a mutable collection inside the loop.

mkeskells commented:
> @mkeskells I've reworked this in what I believe is the best approach for 2.12.x and added benchmark results. Care to take a second look?

Will have a look in the next couple of days

mkeskells commented on src/library/scala/collection/TraversableLike.scala:
```diff
@@ -451,17 +451,23 @@ trait TraversableLike[+A, +Repr] extends Any
-    for (elem <- this) {
-      val key = f(elem)
-      val bldr = m.getOrElseUpdate(key, newBuilder)
+    object m extends mutable.HashMap[K, Builder[A, Repr]] {
```
another good target for `@eager object`

mkeskells replied:
would it not be better to expose the functionality of `entriesIterator` rather than subclassing?
this is effectively a final method in the JVM ATM, and subclassing it here makes all the other calls slightly more expensive

so change this to 
`val m = new mutable.HashMap[K, Builder[A, Repr]]()`
and add 
` @inline private[scala] def entriesIteratorPrivate = entriesIterator` to HashTable.scala

mkeskells commented on src/library/scala/collection/TraversableLike.scala:
```diff
@@ -451,17 +451,23 @@ trait TraversableLike[+A, +Repr] extends Any
+    val m1 = if (m.size > 4) immutable.HashMap.newBuilder[K, Repr] else immutable.Map.newBuilder[K, Repr]
+    while (it.hasNext) {
+      val entry = it.next()
+      m1.+=((entry.key, entry.value.result()))
```
was this intended to call the new API on HashMap `private[collection] def +=(elem: (A, B), keyHashCode: Int)` ?

mkeskells commented on src/library/scala/collection/TraversableLike.scala:
```diff
@@ -451,17 +451,23 @@ trait TraversableLike[+A, +Repr] extends Any
   }
 
   def groupBy[K](f: A => K): immutable.Map[K, Repr] = {
```
I think that we can do better for small collections, so focussing on reducing the allocation cost for small collections and groups , and special casing LinearSeq I come up with something like
```
 def groupBy[K](f: A => K): immutable.Map[K, Repr] = {
    this.seq match {
      case //things with sensible isEmpty and "common" type
        _: LinearSeq[A] |
        _: IndexedSeq[A] |
        _: mutable.Set[A] |
        _: immutable.Set[A]
        if (isEmpty) =>
        //fast path, and avoid additional object creation
        Map.empty
      case _ =>
        val m = new mutable.HashMap[K, Builder[A, Repr]]()
        object grouper extends AbstractFunction1[A, Unit] with Function0[Builder[A, Repr]] {
          override def apply(): mutable.Builder[A, Repr] = newBuilder
          override def apply(elem: A): Unit =
            (m.getOrElseUpdate(f(elem), apply)) += elem
          @tailrec def applyAll(elems: LinearSeq[A]): Unit = {
            if (!elems.isEmpty) {
              apply(elems.head)
              applyAll(elems.tail)
            }
          }
        }
        this.seq match {
          case ls: LinearSeq[A] =>
            grouper.applyAll(ls)
          case _                =>
            this foreach grouper
        }
        m.size match {
          case 0 => Map.empty
          case 1 =>
            //coule be optimised
            val e1 = m.entriesIteratorPrivate.next()
            new immutable.Map.Map1(e1.key, e1.value.result())
          case 2 =>
            val it = m.entriesIteratorPrivate
            val e1 = it.next()
            val e2 = it.next()
            new immutable.Map.Map2( //
                                    e1.key, e1.value.result(), //
                                    e2.key, e2.value.result()
                                    )
          case 3 =>
            val it = m.entriesIteratorPrivate
            val e1 = it.next()
            val e2 = it.next()
            val e3 = it.next()
            new immutable.Map.Map3( //
                                    e1.key, e1.value.result(), //
                                    e2.key, e2.value.result(), //
                                    e3.key, e3.value.result()
                                    )
          case 4 =>
            val it = m.entriesIteratorPrivate
            val e1 = it.next()
            val e2 = it.next()
            val e3 = it.next()
            val e4 = it.next()
            new immutable.Map.Map4( //
                                    e1.key, e1.value.result(), //
                                    e2.key, e2.value.result(), //
                                    e3.key, e3.value.result(), //
                                    e4.key, e4.value.result()
                                    )
//maybe
//          case _ =>
//            val it = m.entriesIteratorPrivate
//            val m1 = immutable.HashMap.newBuilderPrivate[K, Repr]
//            while (it.hasNext) {
//              val entry = it.next()
//              m1.add(entry.key, entry.value.result())
//            }
//            m1.result()
            //but probably better
          case _ =>
            val m1 = immutable.HashMap.newBuilderPrivate[K, Repr]
            m.foreachEntryPrivate(entry => m1.add(entry.key, entry.value.result()))
            m1.result()
        }  }
```
the "maybe" still creates an additional Tuple2 to add to the target HashMap, and recalculates the hashcode of the key
The first part of this is reasonably easy to get around as in "probably better", but I cant see how the to get at the hash from the `HashEntry` in the `mutable.HashMap`. In fact it seems that  `mutable.HashMap` entries should contain a hashcode to reduce the calls to equals, like `java.util.HashMap.Node` does, but that seems outside the scope of this PR for sure

mkeskells replied:
pushed the code or this and the addOne rework for tuples to https://github.com/mkeskells/scala/tree/topic/group-by-2.12.x - just the last commit is the change

retronym committed:
Further performance improvements to groupBy

  - Special case result sizes of 0-4 and avoid mutable.HashMap
    allocation.
  - Expose entriesIterator through a private[collection] method,
    rather than via subclassing, to avoid changing the JIT
    of code using mutable.HashMap that might currently rely on
    Class Heirarchy Analysis noticing that it effectively final.
  - Fuse some lambdas into `object grouper`.

retronym commented:
@mkeskells I've taken most of your ideas into the last commit. Benchmarking with a resulting Map size of 4 to test the special cases for low arity do show an improvement over the previous commit.

```
./benchdb list --limit 3 && ./benchdb results --run 9 --run 8 --run 10 --pivot size  && ./benchdb results --run 9 --run 8 --run 10  --pivot size --metric "·gc.alloc.rate.norm"
┏━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ ID ┃ Timestamp           ┃ Msg                    ┃
┣━━━━╋━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━━━━━━━━━┫
┃ 10 ┃ 2020-08-26 01:11:21 ┃ PR 8948 low-arity-spec ┃
┃  9 ┃ 2020-08-26 00:57:02 ┃ PR 8948 baseline       ┃
┃  8 ┃ 2020-08-26 00:54:20 ┃ PR 8948 commit 1       ┃
┗━━━━┻━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━┛
3 test runs found (limit reached).
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┓
┃                                                 (size) ┃                ┃                ┃      ┃     ┃        128        ┃         512         ┃         2048         ┃          8192          ┃       ┃
┃ Benchmark                                              ┃ (maxNumGroups) ┃ (hashCodeCost) ┃ Mode ┃ Cnt ┃    Score ┃  Error ┃    Score ┃    Error ┃     Score ┃    Error ┃      Score ┃     Error ┃ Units ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━╋━━━━━╋━━━━━━━━━━╋━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━┫
┃ #10:scala.collection.GroupByBenchmark.buildArrayBuffer ┃              4 ┃              0 ┃ avgt ┃  10 ┃ 2172.751 ┃ 61.167 ┃ 7924.906 ┃  246.467 ┃ 31495.987 ┃ 1699.398 ┃ 117165.670 ┃ 11777.052 ┃ ns/op ┃
┃ #8:scala.collection.GroupByBenchmark.buildArrayBuffer  ┃              4 ┃              0 ┃ avgt ┃  10 ┃ 2480.818 ┃ 79.227 ┃ 8334.703 ┃  202.357 ┃ 33122.901 ┃ 1053.086 ┃ 131411.426 ┃  8527.188 ┃ ns/op ┃
┃ #9:scala.collection.GroupByBenchmark.buildArrayBuffer  ┃              4 ┃              0 ┃ avgt ┃  10 ┃ 2589.140 ┃ 70.307 ┃ 9306.434 ┃ 1605.533 ┃ 34211.145 ┃ 2096.226 ┃ 134961.457 ┃  3355.237 ┃ ns/op ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━┻━━━━━━━━━━┻━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━━┻━━━━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━┻━━━━━━━┛
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┓
┃                                                 (size) ┃                ┃                ┃      ┃     ┃       128        ┃        512        ┃       2048        ┃        8192        ┃       ┃
┃ Benchmark [·gc.alloc.rate.norm]                        ┃ (maxNumGroups) ┃ (hashCodeCost) ┃ Mode ┃ Cnt ┃    Score ┃ Error ┃     Score ┃ Error ┃     Score ┃ Error ┃      Score ┃ Error ┃ Units ┃
┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━━━━━━━━━━━╋━━━━━━╋━━━━━╋━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━╋━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━╋━━━━━━━┫
┃ #10:scala.collection.GroupByBenchmark.buildArrayBuffer ┃              4 ┃              0 ┃ avgt ┃  10 ┃ 1128.004 ┃ 0.013 ┃  4328.013 ┃ 0.046 ┃ 16744.052 ┃ 0.183 ┃  66024.199 ┃ 0.699 ┃ B/op  ┃
┃ #8:scala.collection.GroupByBenchmark.buildArrayBuffer  ┃              4 ┃              0 ┃ avgt ┃  10 ┃ 1520.004 ┃ 0.014 ┃  4720.014 ┃ 0.049 ┃ 17136.055 ┃ 0.192 ┃  66416.215 ┃ 0.746 ┃ B/op  ┃
┃ #9:scala.collection.GroupByBenchmark.buildArrayBuffer  ┃              4 ┃              0 ┃ avgt ┃  10 ┃ 3704.004 ┃ 0.015 ┃ 13048.014 ┃ 0.049 ┃ 50040.057 ┃ 0.201 ┃ 197624.222 ┃ 0.771 ┃ B/op  ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━┻━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━┻━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━┻━━━━━━━┛
```

retronym commented:
/rebuild

dwijnand commented:
Needs a rebase and then can be merged, IMO.
